go-controller/pkg/ovn/multicast.go

```diff
@@ -236,16 +252,150 @@ func (oc *DefaultNetworkController) createDefaultAllowMulticastPolicy() error {
+			// default allow ingress
+			aclT = lportIngress
+			dbIDs = getMcastACLDbIDs("", mcastAllowID, aclT, oc.controllerName)
+		} else if strings.Contains(acl.Match, "(igmp || (ip4.src == ") ||
+			strings.Contains(acl.Match, "(mldv1 || mldv2 || (ip6.src == ") {
+			// ns allow ingress
+			aclT = lportIngress
+			// acl Name can be truncated (max length 64), but k8s namespace is limited to 63 symbols,
+			// therefore it is safe to extract it from the name
+			ns := strings.Split(libovsdbops.GetACLName(acl), "_")[0]
+			dbIDs = getMcastACLDbIDs(ns, mcastAllowID, aclT, oc.controllerName)
+		} else if strings.Contains(acl.Match, "ip4.mcast") ||
+			strings.Contains(acl.Match, "(mldv1 || mldv2 || "+ipv6DynamicMulticastMatch+")") {
+			// ns allow egress
+			aclT = lportEgressAfterLB
+			// acl Name can be truncated (max length 64), but k8s namespace is limited to 63 symbols,
+			// therefore it is safe to extract it from the name
+			ns := strings.Split(libovsdbops.GetACLName(acl), "_")[0]
+			dbIDs = getMcastACLDbIDs(ns, mcastAllowID, aclT, oc.controllerName)
+		} else {
+			continue
+		}
+		logLevels := getACLLogLevels(acl)
+		// Update ExternalIDs and Name based on new dbIDs
+		updatedACL := BuildACLFromDbIDs(dbIDs, acl.Priority, acl.Match, acl.Action, logLevels, aclT)
+		updatedACL.UUID = acl.UUID
+		updatedACLs = append(updatedACLs, updatedACL)
+	}
+	err = batching.Batch[*nbdb.ACL](aclUpdateBatchSize, updatedACLs, func(batchACLs []*nbdb.ACL) error {
+		return libovsdbops.CreateOrUpdateACLs(oc.nbClient, batchACLs...)
+	})
+	if err != nil {
+		return fmt.Errorf("cannot update stale multicast ACLs: %v", err)
+	}
+	klog.Infof("Updated %d stale multicast ACLs", len(updatedACLs))
+	return nil
+}
+
+// syncNsMulticast finds and deletes stale multicast db entries for namespaces that don't exist anymore
+func (oc *DefaultNetworkController) syncNsMulticast(k8sNamespaces map[string]bool) error {
+	// to find namespaces that have multicast enabled, we need to find corresponding port groups.
+	// since we can't filter multicast port groups specifically, find multicast ACLs, and then find
+	// port groups they are referenced from.
+	predicateIDs := libovsdbops.NewDbObjectIDs(libovsdbops.ACLMulticastNamespace, oc.controllerName, nil)
+	mcastAclPred := libovsdbops.GetPredicate[*nbdb.ACL](predicateIDs, nil)
+	mcastACLs, err := libovsdbops.FindACLsWithPredicate(oc.nbClient, mcastAclPred)
+	if err != nil {
+		return fmt.Errorf("unable to find multicast ACLs for namespaces: %v", err)
+	}
+	if len(mcastACLs) == 0 {
+		return nil
+	}
+
+	mcastAclUUIDs := map[string]bool{}
+	for _, acl := range mcastACLs {
+		mcastAclUUIDs[acl.UUID] = true
+	}
+	pgPred := func(item *nbdb.PortGroup) bool {
+		for _, aclUUID := range item.ACLs {
+			if mcastAclUUIDs[aclUUID] {
+				return true
+			}
+		}
+		return false
+	}
+	mcastPortGroups, err := libovsdbops.FindPortGroupsWithPredicate(oc.nbClient, pgPred)
```

Comment:
note that you can ignore the result here but build a set with the predicate as well